tests/utils/test_security.py

```diff
@@ -0,0 +1,212 @@
+        client_locator = _KeyLocator(pathlib.Path(test_dir), "curve", "client")
+
+        locators = [server_locator, client_locator]
+
+        for locator in locators:
+            assert locator.public_dir.exists()
+            assert locator.private_dir.exists()
+
+
+def test_key_manager_get_existing_keys_only(
+    test_dir: str, monkeypatch: pytest.MonkeyPatch
+) -> None:
+    """Ensure the key manager cannot load keys when
+    directed not to create missing keys."""
+    with monkeypatch.context() as ctx:
+        ctx.setenv("SMARTSIM_KEY_PATH", test_dir)
+
+        cfg = get_config()
+        km = KeyManager(cfg)
+
+        # use create=False to only load pre-existing keys
+        server_keys, client_keys = km.get_keys(create=False)
+
+        assert server_keys.empty
+        assert client_keys.empty
+
+
+def test_key_manager_get_or_create_keys_default(
+    test_dir: str, monkeypatch: pytest.MonkeyPatch
+) -> None:
+    """Ensure the key manager creates keys when none can be loaded"""
+    with monkeypatch.context() as ctx:
+        ctx.setenv("SMARTSIM_KEY_PATH", test_dir)
+
+        cfg = get_config()
+        km = KeyManager(cfg)
+
+        key_set = km.get_keys()
+
+        # public keys are returned by default
+        assert key_set[0].public != b""
+        assert key_set[1].public != b""
+
+        # default behavior will only return public keys
+        assert not key_set[0].private
+        assert not key_set[1].private
+
+
+def test_key_manager_server_context(
+    test_dir: str, monkeypatch: pytest.MonkeyPatch
+) -> None:
+    """Ensure the key manager does not load private client
+    keys when the context is set to server=True"""
+    with monkeypatch.context() as ctx:
+        ctx.setenv("SMARTSIM_KEY_PATH", test_dir)
+
+        cfg = get_config()
+        km = KeyManager(cfg, as_server=True)
+
+        server_keyset, client_keyset = km.get_keys()
+
+        # as_server=True returns pub/priv server keys...
+        assert len(server_keyset.public) > 0
+        assert len(server_keyset.private) > 0
+
+        # as_server=True returns only public client key
+        assert len(client_keyset.public) > 0
+        assert len(client_keyset.private) == 0
+
+
+def test_key_manager_client_context(
+    test_dir: str, monkeypatch: pytest.MonkeyPatch
+) -> None:
+    """Ensure the key manager does not load private client
+    keys when the context is set to client=True"""
+    with monkeypatch.context() as ctx:
+        ctx.setenv("SMARTSIM_KEY_PATH", test_dir)
+
+        cfg = get_config()
+        km = KeyManager(cfg, as_client=True)
+
+        server_keyset, client_keyset = km.get_keys()
+
+        # as_client=True returns pub/priv client keys...
+        assert len(server_keyset.public) > 0
+        assert len(server_keyset.private) == 0
+
+        # e=True returns only public server key
+        assert len(client_keyset.public) > 0
+        assert len(client_keyset.private) > 0
+
```

Comment:
I'm not sure that I saw a test for the key manager reading keys that already exist. Can we add something like this to the test suite?
```py
def test_key_manager_is_able_to_read_existing_keys_it_created():
    test_dir: str, monkeypatch: pytest.MonkeyPatch
) -> None:
    monkeypatch.setenv("SMARTSIM_KEY_PATH", test_dir)

    cfg = get_config()
    km = KeyManager(cfg, as_client=True, as_server=True)

    # Create some keys
    created_server_keys, created_client_keys = km.get_keys(create=True)
    # ... and then read them
    read_server_keys, read_client_keys = km.get_keys(create=False)
    # ... and check that they are the same
    assert created_server_keys == read_server_keys
    assert created_client_keys == read_client_keys
```

Reply:
renamed `test_key_manager_get_existing_keys_only` to  `test_key_manager_get_existing_keys_only_no_keys_found` and added `test_key_manager_get_existing_keys_only_existing` to test this condition. Great suggestion!